File: src/valueindex/quant.py
"""Advanced-statistics helpers: kernel quantile regression, PCA composite,
AR(1) mean-reversion half-life, and block-bootstrap DCA simulation.

Pure numpy/pandas functions, no ML dependencies, all unit-tested.
"""
from __future__ import annotations

from dataclasses import dataclass

import numpy as np
import pandas as pd

# ---------------------------------------------------------------------------
# A. Conditional quantiles (Nadaraya-Watson-weighted quantile regression)
# ---------------------------------------------------------------------------


def weighted_quantile(values: np.ndarray, weights: np.ndarray, q: float) -> float:
    """Quantile of `values` under observation `weights` (q in [0, 1])."""
    order = np.argsort(values)
    v, w = values[order], weights[order]
    cum = np.cumsum(w) - 0.5 * w
    cum /= w.sum()
    return float(np.interp(q, cum, v))


def conditional_quantiles(
    x: pd.Series,
    y: pd.Series,
    quantiles: tuple[float, ...] = (0.05, 0.25, 0.5, 0.75, 0.95),
    n_grid: int = 60,
    bandwidth: float | None = None,
) -> pd.DataFrame:
    """Kernel-weighted conditional quantiles of y given x.

    For each grid point x0, observations are weighted by a Gaussian kernel
    in x, and weighted quantiles of y are computed. Returns a frame indexed
    by the x grid with one column per quantile (e.g. "q50").
    """
    df = pd.DataFrame({"x": x, "y": y}).dropna()
    xv, yv = df["x"].to_numpy(float), df["y"].to_numpy(float)
    if bandwidth is None:
        # Silverman-flavored rule of thumb, widened for stability in tails.
        bandwidth = 1.06 * xv.std() * len(xv) ** (-1 / 5) * 1.5
    grid = np.linspace(xv.min(), xv.max(), n_grid)
    out = {f"q{int(q * 100)}": np.empty(n_grid) for q in quantiles}
    for i, x0 in enumerate(grid):
        w = np.exp(-0.5 * ((xv - x0) / bandwidth) ** 2)
        # Skip grid points with too little effective data.
        if w.sum() < 5:
            for q in quantiles:
                out[f"q{int(q * 100)}"][i] = np.nan
            continue
        for q in quantiles:
            out[f"q{int(q * 100)}"][i] = weighted_quantile(yv, w, q)
    return pd.DataFrame(out, index=pd.Index(grid, name="x"))


# ---------------------------------------------------------------------------
# B. PCA composite valuation index
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class PcaComposite:
    composite: pd.Series          # first-PC score, z-scaled, monthly
    weights: pd.Series            # loading per indicator (sums of squares = 1)
    explained_ratio: float        # share of total variance carried by PC1


def pca_composite(aligned_z: pd.DataFrame, min_series: int = 4) -> PcaComposite:
    """First principal component of direction-aligned z-scored indicators.

    Indicators start at different dates, so the correlation matrix uses
    pairwise-complete observations and per-row scores use only the
    indicators available that month (weights renormalized). Rows with
    fewer than `min_series` indicators are dropped.
    """
    corr = aligned_z.corr(min_periods=24)
    eigvals, eigvecs = np.linalg.eigh(corr.to_numpy())
    v1 = eigvecs[:, -1]
    if v1.sum() < 0:  # sign convention: composite up = more expensive
        v1 = -v1
    weights = pd.Series(v1, index=aligned_z.columns)
    explained = float(eigvals[-1] / eigvals.sum())

    z = aligned_z.to_numpy(float)
    w = weights.to_numpy(float)
    mask = ~np.isnan(z)
    eff_w = np.where(mask, w, 0.0)
    norm = np.sqrt((eff_w**2).sum(axis=1))
    score = np.nansum(z * eff_w, axis=1) / np.where(norm == 0, np.nan, norm)
    score[mask.sum(axis=1) < min_series] = np.nan
    composite = pd.Series(score, index=aligned_z.index).dropna()
    composite = (composite - composite.mean()) / composite.std()
    return PcaComposite(composite=composite, weights=weights, explained_ratio=explained)


# ---------------------------------------------------------------------------
# Nearest historical analogs (z-space nearest-neighbor search)
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class Analog:
    date: pd.Timestamp
    distance: float      # RMS difference per shared indicator, in sigmas
    shared: int          # number of indicators available in both months


def nearest_analogs(
    aligned_z: pd.DataFrame,
    n: int = 4,
    min_separation_months: int = 36,
    exclude_recent_months: int = 60,
    min_shared: int = 4,
) -> list[Analog]:
    """Historical months most similar to the latest month in indicator space.

    Distance is the RMS z-score difference over indicators present in BOTH
    months (indicators start at different dates). The most recent
    `exclude_recent_months` are excluded (trivially similar), and picked
    analogs must be `min_separation_months` apart so one episode doesn't
    fill every slot.
    """
    target = aligned_z.iloc[-1]
    hist = aligned_z.iloc[:-1]
    if exclude_recent_months:
        hist = hist.iloc[:-exclude_recent_months]
    if hist.empty:
        return []

    diff_sq = (hist - target) ** 2
    shared = diff_sq.notna().sum(axis=1)
    dist = np.sqrt(diff_sq.mean(axis=1, skipna=True))
    dist[shared < min_shared] = np.nan

    picked: list[Analog] = []
    for date, d in dist.dropna().sort_values().items():
        if any(abs((date - p.date).days) < min_separation_months * 30 for p in picked):
            continue
        picked.append(Analog(date=date, distance=float(d), shared=int(shared[date])))
        if len(picked) >= n:
            break
    return picked


# ---------------------------------------------------------------------------
# C. AR(1) mean-reversion half-life
# ---------------------------------------------------------------------------


def ar1_half_life_months(s: pd.Series) -> float | None:
    """Half-life (months) of deviations from the mean under an AR(1) fit.

    phi is the lag-1 autocorrelation of the demeaned series; half-life is
    ln(0.5)/ln(phi). Returns None when the series doesn't mean-revert
    (phi >= 1) or is too short.
    """
    s = s.dropna()
    if len(s) < 36:
        return None
    x = s - s.mean()
    phi = float(x.autocorr(lag=1))
    if not (0 < phi < 1):
        return None
    return float(np.log(0.5) / np.log(phi))


# ---------------------------------------------------------------------------
# D. Block-bootstrap DCA (monthly contribution) simulation
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class DcaResult:
    percentiles: pd.DataFrame     # index: month, columns p5/p25/p50/p75/p95
    final_wealth: np.ndarray      # per-simulation final wealth
    total_contributed: float


def simulate_dca(
    monthly_returns: pd.Series,
    monthly_contribution: float,
    years: int,
    n_sims: int = 2000,
    block: int = 12,
    annual_drift_target: float | None = None,
    seed: int = 42,
) -> DcaResult:
    """Simulate dollar-cost averaging with block-bootstrapped returns.

    Contiguous `block`-month chunks of history are resampled to preserve
    volatility clustering. If `annual_drift_target` (percent, real) is
    given, the resampled returns are shifted so their expected annual
    return matches it — this plugs the Bogle-calculator estimate in while
    keeping historical volatility.
    """
    rng = np.random.default_rng(seed)
    rets = monthly_returns.dropna().to_numpy(float)
    horizon = years * 12
    n_blocks = int(np.ceil(horizon / block))
    starts = rng.integers(0, len(rets) - block, size=(n_sims, n_blocks))
    idx = (starts[:, :, None] + np.arange(block)[None, None, :]).reshape(n_sims, -1)
    sims = rets[idx[:, :horizon]]

    if annual_drift_target is not None:
        target_monthly = (1 + annual_drift_target / 100) ** (1 / 12) - 1
        sims = sims + (target_monthly - rets.mean())

    wealth = np.zeros((n_sims, horizon))
    w = np.zeros(n_sims)
    for t in range(horizon):
        w = (w + monthly_contribution) * (1 + sims[:, t])
        wealth[:, t] = w

    pct = pd.DataFrame(
        {
            f"p{p}": np.percentile(wealth, p, axis=0)
            for p in (5, 25, 50, 75, 95)
        },
        index=pd.RangeIndex(1, horizon + 1, name="month"),
    )
    return DcaResult(
        percentiles=pct,
        final_wealth=wealth[:, -1],
        total_contributed=monthly_contribution * horizon,
    )

File: docs/assets/js/quant.js
// Port of valueindex/quant.py: PCA composite (Jacobi eigensolver on the
// pairwise correlation matrix) and nearest-analog search. Semantics match
// the Python implementation exactly (see tests/e2e parity fixtures).

import { asofIndex, mean, std } from "./stats.js";

// Direction-aligned z-score columns on the panel's shared grid, optionally
// truncated at `asof` (the time machine). Mirrors app.py's aligned_z_panel.
export function alignedZFromPanel(panel, registry, asof = null) {
  const cut = asof === null ? panel.dates.length - 1 : asofIndex(panel.dates, asof);
  const dates = panel.dates.slice(0, cut + 1);
  const columns = {};
  for (const key of registry.valuation_keys) {
    const sign = registry.indicators[key].higher_is_expensive ? 1 : -1;
    const raw = panel.series[key].slice(0, cut + 1);
    const vals = raw.filter((x) => x !== null);
    if (vals.length < 2) {
      columns[key] = raw.map(() => null);
      continue;
    }
    const m = mean(vals), s = std(vals);
    columns[key] = raw.map((x) => (x === null ? null : ((x - m) / s) * sign));
  }
  return { dates, columns };
}

// ---- pairwise-complete Pearson correlation (pandas corr(min_periods)) ----

function pearson(a, b) {
  // over indices where both non-null; returns [corr, count]
  let n = 0, sa = 0, sb = 0;
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== null && b[i] !== null) { n++; sa += a[i]; sb += b[i]; }
  }
  if (n < 2) return [NaN, n];
  const ma = sa / n, mb = sb / n;
  let cov = 0, va = 0, vb = 0;
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== null && b[i] !== null) {
      const da = a[i] - ma, db = b[i] - mb;
      cov += da * db; va += da * da; vb += db * db;
    }
  }
  const denom = Math.sqrt(va * vb);
  return [denom === 0 ? NaN : cov / denom, n];
}

// ---- Jacobi eigendecomposition for a symmetric matrix ----

export function jacobiEigh(matrix) {
  const n = matrix.length;
  const a = matrix.map((row) => row.slice());
  let v = Array.from({ length: n }, (_, i) =>
    Array.from({ length: n }, (_, j) => (i === j ? 1 : 0))
  );
  for (let sweep = 0; sweep < 200; sweep++) {
    let off = 0;
    for (let p = 0; p < n; p++)
      for (let q = p + 1; q < n; q++) off += a[p][q] * a[p][q];
    if (off < 1e-22) break;
    for (let p = 0; p < n; p++) {
      for (let q = p + 1; q < n; q++) {
        if (Math.abs(a[p][q]) < 1e-15) continue;
        const theta = (a[q][q] - a[p][p]) / (2 * a[p][q]);
        const t = Math.sign(theta || 1) / (Math.abs(theta) + Math.sqrt(theta * theta + 1));
        const c = 1 / Math.sqrt(t * t + 1);
        const s = t * c;
        for (let k = 0; k < n; k++) {
          const akp = a[k][p], akq = a[k][q];
          a[k][p] = c * akp - s * akq;
          a[k][q] = s * akp + c * akq;
        }
        for (let k = 0; k < n; k++) {
          const apk = a[p][k], aqk = a[q][k];
          a[p][k] = c * apk - s * aqk;
          a[q][k] = s * apk + c * aqk;
        }
        for (let k = 0; k < n; k++) {
          const vkp = v[k][p], vkq = v[k][q];
          v[k][p] = c * vkp - s * vkq;
          v[k][q] = s * vkp + c * vkq;
        }
      }
    }
  }
  const eigvals = a.map((row, i) => row[i]);
  // eigenvectors are columns of v
  return { eigvals, eigvecs: v };
}

// ---- PCA composite (port of quant.pca_composite) ----

// columns: {key: array-with-nulls on a shared grid}
export function pcaComposite(keys, columns, minSeries = 4, minObs = 24) {
  const usable = keys.filter(
    (k) => columns[k].reduce((c, x) => c + (x !== null ? 1 : 0), 0) >= minObs
  );
  minSeries = Math.min(minSeries, Math.max(1, usable.length));

  const gridLen = usable.length ? columns[usable[0]].length : 0;

  if (usable.length === 1) {
    const raw = columns[usable[0]];
    const vals = raw.filter((x) => x !== null);
    const m = mean(vals), s = std(vals);
    const composite = raw.map((x) => (x === null ? null : (x - m) / s));
    return { usable, weights: { [usable[0]]: 1.0 }, explained: 1.0, composite };
  }

  const n = usable.length;
  const corr = Array.from({ length: n }, () => new Array(n).fill(0));
  for (let i = 0; i < n; i++) {
    corr[i][i] = 1;
    for (let j = i + 1; j < n; j++) {
      const [r, count] = pearson(columns[usable[i]], columns[usable[j]]);
      const val = count >= minObs && !Number.isNaN(r) ? r : 0; // fillna(0)
      corr[i][j] = val;
      corr[j][i] = val;
    }
  }

  const { eigvals, eigvecs } = jacobiEigh(corr);
  let maxI = 0;
  for (let i = 1; i < n; i++) if (eigvals[i] > eigvals[maxI]) maxI = i;
  let v1 = eigvecs.map((row) => row[maxI]);
  if (v1.reduce((a, b) => a + b, 0) < 0) v1 = v1.map((x) => -x);
  const explained = eigvals[maxI] / eigvals.reduce((a, b) => a + b, 0);
  const weights = Object.fromEntries(usable.map((k, i) => [k, v1[i]]));

  // per-row score with renormalized weights over available indicators
  const raw = new Array(gridLen).fill(null);
  for (let r = 0; r < gridLen; r++) {
    let dot = 0, norm2 = 0, present = 0;
    for (let i = 0; i < n; i++) {
      const x = columns[usable[i]][r];
      if (x !== null) { dot += x * v1[i]; norm2 += v1[i] * v1[i]; present++; }
    }
    if (present >= minSeries && norm2 > 0) raw[r] = dot / Math.sqrt(norm2);
  }
  const vals = raw.filter((x) => x !== null);
  const m = mean(vals), s = std(vals);
  const composite = raw.map((x) => (x === null ? null : (x - m) / s));
  return { usable, weights, explained, composite };
}

// ---- nearest analogs (port of quant.nearest_analogs) ----

const DAY = 86400000;

function daysBetween(d1, d2) {
  return Math.abs(Date.parse(d1) - Date.parse(d2)) / DAY;
}

// columns: {key: array-with-nulls}, dates: shared grid (ascending)
export function nearestAnalogs(keys, columns, dates, opts = {}) {
  const { n = 4, minSepMonths = 36, excludeRecentMonths = 60, minShared = 4 } = opts;
  const last = dates.length - 1;
  const target = keys.map((k) => columns[k][last]);
  const histEnd = last - excludeRecentMonths; // exclusive of target row already
  const candidates = [];
  for (let r = 0; r < histEnd; r++) {
    let sum = 0, shared = 0;
    for (let i = 0; i < keys.length; i++) {
      const a = columns[keys[i]][r], b = target[i];
      if (a !== null && b !== null) { const d = a - b; sum += d * d; shared++; }
    }
    if (shared >= minShared) {
      candidates.push({ date: dates[r], distance: Math.sqrt(sum / shared), shared });
    }
  }
  candidates.sort((a, b) => a.distance - b.distance);
  const picked = [];
  for (const c of candidates) {
    if (picked.some((p) => daysBetween(p.date, c.date) < minSepMonths * 30)) continue;
    picked.push(c);
    if (picked.length >= n) break;
  }
  return picked;
}

export function monthsBetween(d1, d2) {
  const [y1, m1] = [Number(d1.slice(0, 4)), Number(d1.slice(5, 7))];
  const [y2, m2] = [Number(d2.slice(0, 4)), Number(d2.slice(5, 7))];
  return (y2 - y1) * 12 + (m2 - m1);
}
